### ISSUE TITLE
Tracing an incorrect batch statement may cause segmentation fault

### ISSUE THREAD

**Issue by psarna** [scylladb/scylla#9220]
The code responsible for producing tracing info makes certain assumptions about its input - that the number of recorded prepared statements is the same as the number of recorded names and values for each of these statements. This assumption holds true only if all statements in a batch are correct. That happens because the vector of prepared statements in tracing is updated separately from the vector of its matching names and values, and it's possible to throw between these two updates, which in turns leads to a segmentation fault.

A fix is already on its way. Easy to backport and important to do so, because an incorrect batch combined with probabilistic tracing may kill a server.

**Comment by vladzcloudius** [comment]
The issue is regression introduced by 

```
commit 4893bc9139cab4b3627693284a22d4a647dc3092
Author: Gleb Natapov <gleb@scylladb.com>
Date:   Thu Dec 5 12:20:26 2019 +0200

    tracing: split adding prepared query parameters from stopping of a trace
    
    Currently query_options objects is passed to a trace stopping function
    which makes it mandatory to make them alive until the end of the
    query. The reason for that is to add prepared statement parameters to
    the trace.  All other query options that we want to put in the trace are
    copied into trace_state::params_values, so lets copy prepared statement
    parameters there too. Trace enabled case will become a little bit more
    expensive but on the other hand we can drop a continuation that holds
    query_options object alive from a fast path. It is safe to drop the call
    to stop_foreground_prepared() here since The tracing will be stopped
    in process_request_one().
    
    Message-Id: <20191205102026.GJ9084@scylladb.com>

```

**Comment by vladzcloudius** [comment]
Introduced in 3.3.0

**Comment by vladzcloudius** [comment]
The description of the problem is somewhat misleading.
The issue can be reproduced not only with BATCHes but with non-BATCH statements too. All it takes is to send an EXECUTE with a wrong number of parameters and trace such an EXECUTE.

Another option to hit this is if `emplace_back` throws here: https://github.com/scylladb/scylla/blob/scylla-4.5.rc5/tracing/trace_state.cc#L160.

So, the exposure is much wider than in the description above.
@eliransin @amihayday FYI

### Now let's describe how we have got to this in a bit more detail:

The assumptions in the tracing code belong to the code that has been written prior to regression where those assumptions actually held because a pointer to `options` was passed to the `tracing::build_parameters_map(options_ptr)` if and only if corresponding checks in upper layers had passed.

In case of any error nullptr was passed and no dereferencing was taking place at all: https://github.com/scylladb/scylla/blob/scylla-3.2.0/tracing/trace_state.cc#L118

Before 4893bc9139cab4b3627693284a22d4a647dc3092 the only array `trace_state` had was the one that held prepared statements and we used other two arrays from the passed `options` object and we used safe accessors for fetching items from it but more importantly `options` was promised to be properly validated before we would even attempt to dereference it.

So, with the old code we were absolutely safe.

4893bc9139cab4b3627693284a22d4a647dc3092 introduced two new arrays in `trace_state` and used unsafe accessors and this can break as described above.

### As to an idea for a fix:

https://github.com/scylladb/scylla/pull/9221 is going to work the issue around but would leave the code in a spaghetti state.

There was a single array in the original code (before 4893bc9139cab4b3627693284a22d4a647dc3092). 

Having 3 arrays that need to be in sync and having to check this every time we need to access them is a very bad design pattern.

Instead we should store all 3 corresponding data items in a single object and have a single array that holds such triplets.

**Comment by eliransin** [comment]
fixed by  ec52e05 which AFAIU implements @vladzcloudius sugetion above (please reopen if I am wrong).
